### ISSUE TITLE
Conditional reference inclusion causes references to disappear from Solution Explorer

### ISSUE THREAD

**Issue by drvink** [dotnet/fsharp#329]
[This issue was originally posted on the VS Connect tracker.](https://connect.microsoft.com/VisualStudio/feedback/details/1037540/conditional-reference-inclusion-causes-references-to-disappear-from-solution-explorer) Lincoln Atkinson confirmed there was an issue and requested that it be moved here for better visibility.
#### Summary

When using conditional compilation to include/exclude certain features, one might also want to make conditional the referencing of an assembly or linking of a library. This isn't difficult to do with MSBuild, but it seems to interact poorly with Visual Studio. Please see "steps to reproduce" for directions, but first, take a look at libhomie\Libhomie\Libhomie.fsproj. The important lines are line 2 (the Project element), lines 18 and 19 (the Import element), the two lines with DefineConstants, and lines 41 to 57. Also see libhomie\Libhomie.targets. This is the conditional assembly referencing stuff.

It seems you can't attach files to GitHub issues other than images. [Here is the attachment from the original ticket](http://triplehelix.org/~mdl/tmp/conditional_reference_bug.zip).
#### Steps to reproduce
1. Open User_of_libhomie\User_of_libhomie.sln in Visual Studio.
2. Notice that there is no disclosure triangle next to "References" for Libhomie in the Solution Explorer. (This is the bug!)
3. Open Program.fs in VS from the Solution Explorer.
4. Build the solution to make the red wavy lines go away. Try running the program.
5. Open Seq.fs in VS from the Solution Explorer. Notice the conditional stuff between #if WANT_EXTCORE/#endif.
6. Go to User_of_libhomie\User_of_libhomie\bin\$(Configuration) in Windows Explorer, XTree Gold, Vernon Buerg's list.com, whatever. Notice that ExtCore was (correctly) not part of the dependency chain.
7. Clean the solution and close it.
8. Open Libhomie\Libhomie.sln in Visual Studio.
9. Open Seq.fs in VS. Notice that this time, the conditional compilation stuff is enabled (which is fine, that's what we want).
10. Build the solution.
11. Go to libhomie\Libhomie\bin\$(Configuration) in your favorite file guy. Notice that the ExtCore assembly was copied in (which is correct, since the default configuration of Libhomie is to include features that depend on ExtCore).
#### Actual results

References in the Solution Explorer are missing. Also, as noted in Libhomie.fsproj, there's no workaround--if we make the AddExtCoreReference target anything other than an InitialTarget(s) of the project, the build will break.
#### Expected results

References should be visible.

(The following comment was because the original ticket was filed under the VS Connect bug tracker. Now that this issue is on the Visual F# tracker, I realize that changing it is not up to only the Visual F# team, though I think it would improve the VS build experience as a whole.)

It seems like a major oversight that the Visual Studio IDE has no way (short of writing a very, very complicated VSPackage) of simply providing additional properties the way that the /p option to msbuild.exe does.

Are there any plans to make it so that solutions are just MSBuild project files? The fact that MSBuildEmitSolution exists is nice, but it doesn't solve the fundamental problem of solutions being separate, "non-programmable" entities. For example, if a solution were just a project file, the implementation of what I've demonstrated here could be made far simpler ([you could do it via AdditionalProperties as mentioned on this page](http://msdn.microsoft.com/en-us/library/z7f65y0d.aspx)).

**Comment by drvink** [comment]
Another comment from the original ticket:

This same issue occurs when attempting to set the TargetFrameworkVersion property; Visual Studio does not seem to resolve the property correctly, assuming the project is targeting the same version regardless of configuration (after the version is intially resolved when loading the project).

**Comment by bentayloruk** [comment]
@drvink with regards to the `TargetFrameworkVersion` Visual Studio issues, it *seems* that this is a known VS issue and one that won't be fixed.  I ran into this yesterday, after two days of trying to convert a 10 project solution to use a conditional framework version!  The Connect bugs containing the relevant information are linked off [this Stack Overflow answer](http://stackoverflow.com/a/23790305/418492).  I appreciated that this might be fixable for F# projects, but it looks like it could be inconsistent with the "official" VS advice.
